src/parser/shared/metrics/buffApplications.ts

```diff
@@ -0,0 +1,82 @@
+import { AnyEvent, EventType } from 'parser/core/Events';
+import metric from 'parser/core/metric';
+
+/**
+ * Returns the index of the last element in the array where predicate is true, and -1
+ * otherwise.
+ * Source: https://stackoverflow.com/a/53187807/684353
+ * @param array The source array to search in
+ * @param predicate find calls predicate once for each element of the array, in descending
+ * order, until it finds one where predicate returns true. If such an element is found,
+ * findLast immediately returns that element index. Otherwise, findLastIndex returns undefined.
+ */
+export function findLast<T>(
+  array: T[],
+  predicate: (value: T, index: number, obj: T[]) => boolean,
+): T | undefined {
+  let l = array.length;
+  while (l !== 0) {
+    l -= 1;
+    if (predicate(array[l], l, array)) {
+      return array[l];
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Returns an object with all buff applications. Should probably not be used
+ * directly in configs, but only internally by other shared stats.
+ */
+const buffApplications = (events: AnyEvent[]) =>
+  events.reduce<{
+    [spellId: number]: {
+      [sourceId: number]: Array<{
```

Comment:
What's the motivation behind indexing by spellId and sourceId but not by target? It seems like it would be much easier to detect buff times per target if we indexed by target. Also, won't almost everything have sourceId === the player?

Reply:
target also has a targetinstance, which can vary but targetId can be repeated. If the player is affected by someone else's buff, like Bloodlust, the source will be someone else. There are quite a few buffs like that.

Reply:
I know about targetInstance, in existing code I try to abstract that out with `encodeTargetString`, but some code forgets to use it. I fixed a bug with that just last week. Would you consider using something like `encodeTargetString` to index this object?

Reply:
It can be just as easy to detect buff times with this array if we provide a selector function. What remains is a performance question, which might be too early.

I was hesitant to use `encodeTargetString` because it would add complexity while this solution works as well. I think it makes sense to go source -> spell -> target since that reflects how they're cast (player casts spell on target). We're also pretty likely to be looking at buffs cast *by* the player, though perhaps then we'd look at the cast event instead. Maybe it should be changed around after all.

Reply:
Will work on that later.